### DISCUSSION TITLE
can vanilla-extract work with fresh

### DISCUSSION THREAD

**Discussion by dagda1** [vanilla-extract-css/vanilla-extract#960] [Q&A]
I really like the look of the Deno and the [fresh framework](https://fresh.deno.dev/), and I'm interested to know if vanilla-extract can run on the fresh framework.

Fresh provides [plugins](https://fresh.deno.dev/docs/concepts/plugins), but I'm not sure if they can be used with vanilla-extract.

At the moment, a fresh plugin has one hook, `render`, and there are no build time hooks.

Here is a simple gist of an [emotion]() plugin and here are a few lines of code:

```ts
const emotion: Plugin = {
    name: "emotion",
    entrypoints: { "main": main },
    render: (ctx) => {
        const res = ctx.render()
        const { css, ids } = extractCritical(res.htmlText)
```

This is very much happening at runtime and not build time.

Is it possible to hook vanilla-extract at runtime or is it too late then?

**Comment by jd-oconnor** [comment]


**Reply by dagda1** [reply]
> I could be wrong but from a quick glance at the repo, it looks like deno uses esbuild. You could spin up a test app, build 1 basic component with VE and follow the esbuild setup docs and see if it works
> 
> 
> 
> https://vanilla-extract.style/documentation/integrations/esbuild/

Sorry, I'm not sure I follow.  Which repo are you talking about?  I'm talking about the deno web framework fresh that I linked to

**Reply by dagda1** [reply]
I'm not talking about deno itself

**Reply by jd-oconnor** [reply]
https://github.com/denoland/fresh/blob/main/src/server/bundle.ts

That’s from fresh framework. It uses esbuild. If you make an app using fresh, it’ll use esbuild.

VE has an esbuild integration with steps you can follow so fresh will be able to work with VE
https://vanilla-extract.style/documentation/integrations/esbuild/

**Reply by dagda1** [reply]
@jd-oconnor, you are pointing to the actual build of the framework.  When you create a fresh project, they do not expose esbuild. The way they want you to go down is the plugin (which was my original point before all this).

There is no natural build step in fresh (which I pointed out initially), and my question remains, would it be possible to create a plugin for fresh with its limitations and vanilla-extract.

If you don't know, it is better not to answer because I find it difficult to get answers in VE discord or here, and I fear a wrong answer has derailed this discussion with a quick and unthought-through answer.

Apologies for the reply, but I don't think I am any closer to knowing, and all I've done is type quite a bit on my keyboard

**Reply by jd-oconnor** [reply]
@dagda1 let me quote you in discord

> Could someone please give me an opinion of whether vanilla-extract can be used with the Deno web framework fresh. I created a discussion here

>https://github.com/vanilla-extract-css/vanilla-extract/discussions/960

>Any answer no matter how short appreciated

I responded to you on discord and also put it here. Not only did I admit I could be wrong, but I’m the only person who even attempted to help you and instead of talking to me like a decent person, you responded like a pompous ass. It’s no wonder no one is helping you if this is how you treat them. Clearly you don’t appreciate it.

**Comment by dagda1** [comment]
@jd-oconnor I know, I apologise for my response but I just think no response would have been better here if you did not know.
